### PR TITLE
Fixes some bugs in examples and factorize some code, test them in CI

### PR DESCRIPTION
**Pull request by illwieckz**
Fixes some bugs in examples and factorize some code.

The examples are now tested in CI.

### CODE REVIEW

**Comment by illwieckz** [comment]
The example1 CRN to DDS test doesn't work on macOS CI, I don't know why because I don't reproduce it on my end on macOS.

**Comment by illwieckz** [comment]
```
running: build/example1 c test/unvanquished_64.png -crn -out build/test/9/unvanquished_64.crn
example1 - Version v1.04 Built Jul  8 2024, 01:50:07
Loading source file: test/unvanquished_64.png
Source file size: 7471, Dimensions: 64x64
Actual Components: 4
Compressing to DXT5
Generating mipmaps using filter "kaiser"
Generated 6 mipmap levels in 0.004s
Compressing using quality level 128
Processing: 100%
Compressed to 2210 bytes, quality level: 128, effective bitrate: 3.237502
Writing CRN file: build/test/9/unvanquished_64.crn
running: build/example1 d build/test/9/unvanquished_64.crn -out build/test/9/unvanquished_64.dds
Loading source file: build/test/9/unvanquished_64.crn
Decompression mode only supports .dds or .crn files!
+ base64 build/test/9/unvanquished_64.crn
SHgAYoiYAAAIoquhAEAAQAcBAgAAAAAAAAAAAAEAAAAAAABiAAGMAIsAAe4AAd0ApAADywAAUwAyAAQeAADpADoAogAFBwAABakAAAeTAAAIPQAACHMAAAiCAAAIjwAACJoAgiwBG1BBBBLvN5NlUQhEDS/twBAfJAIgwww0k3JS5mg4GcFDSeW2NoCAcIWhCEInLprVIDn96AJApCAWDrMAUgDxwBa0SurOqUVmKYgUl2guMACROXV2SrsdtULQN0usCkupSc13AJBVLTADE3WwilRWDR2QgBATSLFxmA0KipNu0l6NL3HAtSXddLt1XE6AAsK3Uxl51SO27pK7bL1prtBASZktuqWJiN1v0dHd7fQDN/SC9dP/6+c79y+VdxkznQ/Ortbvz3AdTatCXY7uuxnEtb5E3j5LvRLOCWG12gmMk4YgLXNnC6hoBMhpbCqrVzHR1hEbZhhs9kwEIWYHEQnISGSiWtJSred5JJbOEPQhR3l7yUxKDl6EmXNa/eiFnnv872yueefzvHe6pd+RzQ3jzdvUFqegoEy1drbVK289y9gom1tM225m4d1fVlqxX1+9zfmb8c8/0c9mcmekSEu3bnibRvze0/DaO29v73z1/+b/rm2SCV2CucuZuOLPXf7+cb/PH988fsAAQmAwGwgxAwBOQ15i1kAjoB6oXOOEat4jljeUwub+jEGmgFzgMAwAChkOSqvKrlhJNGZ9ot3w2L51mNIAoPsofMGYQBcMw+Bkygr8oEoHQD0BPwGAPgA7DDYOrAWAjKpviRTlcCdgJrAe7gDoEMwythZooaJkne+0sXub1H9rSCTyb1Tpwh9tKqJcwzAPuGwq9XstmoeqqrKiMm0cox0xyXYynNtlWudgvkzK3ar/VuW8JZJMSXmss86L3NRzPftheAEaaaNtttgme6qhoZjuIEmoqtpwJpI7u+sil8y6emrLxvTZ7p2uWSz8fbzvdPfpOafvL0oZO5yNoiIjcQkExwue04PzLNn/2RDhqdyv8eN0/kd5YmSIDTThxfXr1xhHSyOSOp506k66cVf7uV5NxOGp3O/XJjLW0NVu7ZVXG4pbkRib02jJRTelmIqydecVJ9eHjGD26smToq5gzM+2liZxSal8zeI5S33T1PXsZ2Wa7zQu+nu76fus04uO5barAqxjGM4uW9G1Zxb+bZ9j9Gecu23jN5JprjjJ3ljzmCXpYTMF788Wt9a3OGjreb83WOj9NTUdO9eu+P7jI4j919uvd9fcz2PdZprc2Bof7vr3Xmufs6mtTWvft/wEAjHQggxRRRMzvn4CgiK6EhBGhGjWkW03Y0E0A0E6JIScDmJY9HnbdhYFnE1CIlNQsUVlpM4Ybt/VVWUdNNc5LG3ebG84Nx2LyhEfnsx/f6++IAECdACbTEAAAF93TTu6uovW4ppGUJPhtIYUsRGMrQ1ACYJiaYCKKKIkkkk0aNGiig8N+uzR6uHEC/CxYsGGAE0EBF6PQZ/q1aKPh9xDdy1tBRzBBYBueCCTK0g+zGCgDRQKScA+aLw3bgDDAH8Xh8b8IhyKO7u7v2GktJaS0kBvVVVc2LFilgB1/AceIyZNhkz1kCigRR8042IOmFyytlbK3+3nsqqpugqrr11kzZqAMkUUUbp3TundNCLxZ+Y/3h/q/X2v42Xdunbp6w4fS/xHPrv29+eE6TpHOOd3/297Z2zx5edsCSSQAqXpgMAMMMNgoGsKcZeMYLjwFBOcmyWSfGUmAi20YJIMQbfbuznZQc7VLhGESs7VMzEZDaxalvCgiw+FREZQhIGg1t4RqLQ1dvx4CkeA6ChBgBngJESYj7iZhCE1RTURHMUkTF0TE8DNSaiH5IsJEVIAZTooTQYYoAAvJBFhsFfHfDBaXL2UHTNV6AdTAgBYYYAAAm2FFEcpHkohGxH2sfYAaPER4+Yj8IxBGgCIiyhGgI0SgiIzFmKClI/IiIu0EfbYRvO01CWgbgbgbgbhLwmeM1Dd9H4H4TUDUJ+GjzEdCTfiMBrtSptNByRAVXklkCON+7VZ2lJ1npPPphKlukj00Y2JPNDdrHBS0QyBMXtRaJi6u3rezxeLwPuB/mWQY2YrknmiSRQcoIlL1o+evypYHZhRddtyguboDhfc3CrdX3kaTQyxJxkrSAaoAjXb5WWGMJ7nLc6o4l5zKHk72NEGW6JO7Fag5UI5zwIUVkrTLAXwoubtvEeVBPd5Jpn44QcGWJMkJAGzgeZ4uLh5tFpj/T7rnKEPFDrAcoJkoGWZ1yTTjQGEnKKZVg1R5YhppPjpn6ZIlsD5gdX5IYS1csnQyxJ0YoBKKVJRXKJQuk4SGZ9DHVg6GWY7kn/wHZQDdQLYUaCKprteF1KqREDYBwXC6Wguf3GMMsSZQFAKErYToMkNthlQ41OGOMsR9STr7KUxtoGZSEhGsQrxsXMZPMjACUedsX12NHDgPC2jGxJ59sxkLwFAYi9FYZtEQl8qTkps/is8MbMzWn6GbWPmvf4R19TCuxGFWFdGriqJNWZ5r6pa6QIq/hNXQiaqRFVCrWeUX0g+REc/b2EhISEhISMkSEhISEhIyo6h2It+1+PYQOh7Ccwj9md5J7CBVoMNrlrU9KOaTl5DO+3LP/6vLfuycOy/bSF9KNTI17+lGtYAztcnK3YMveS28vLmYxMto6auVtRDXjKSkvdjG8n39PF0fIOrTK/qLWkOtUm44nwSs+uflpjHnKXX02Mvu+zQKQMA6fLp1mmSUvfUJnt+B05tcyf3g6U2+dLJdpRrP07kNMje1N0oMeDGqq8Qm3xpcaikxQdYC2UpxbMzHBGusrw534PLn0c/vrfD08m1Jvlaw/vp6FufCFmejPH/ZOv5/bT+jXOpdYaoaWfiDv5m70k/3vDX99OLh2+oTUd321kw9sE/f6X9AHyeMXfxyXfibheY38eS6F0LoXQ=
```

If I decode that file on my end and run crunch (even macOS crunch on it), it says it's a valid crn file and decompresses it!

**Comment by illwieckz** [comment]
> The example1 CRN to DDS test doesn't work on macOS CI, I don't know why because I don't reproduce it on my end on macOS.

Now fixed, a null-terminator was missing.

**Comment by illwieckz** [comment]
I don't know why the `example1` test doesn't run on Windows MSVC Appveyor CI , while on my end a MinGW build runs correctly on wine.

**Comment by slipher** [comment]
> I don't know why the `example1` test doesn't run on Windows MSVC Appveyor CI , while on my end a MinGW build runs correctly on wine.

I reproduce a crash. It crashes when calling `splitpath_s`. I don't know what exactly was wrong, but anyway there is already a portable version `file_utils::split_path`. So `example_platform` files shouldn't be needed.

**Comment by illwieckz** [comment]
> I don't know what exactly was wrong, but anyway there is already a portable version `file_utils::split_path`. So `example_platform` files shouldn't be needed.

It should be doable for `example1` (I have a PoC that works in a branch), but that would be more problematic for `example2` that is meant to be a example of how to _not_ rely on crnlib:

> ### example2
>
> Shows how to transcodec CRN files to DDS using **only**
> the functionality in `inc/crn_decomp.h`. It does not link against against
> `crnlib.lib` or depend on it in any way. (Note: The complete source code,
> approx. 4800 lines, to the CRN transcoder is included in `inc/crn_decomp.h`.)

**Comment by illwieckz** [comment]
So `example1` and `example3` use crnlib for both `split_path` and `strcpy_safe`, but `example2` use custom code.

File: test/test.py [hunk 1]
```diff
@@ -4,32 +4,62 @@
 import subprocess
 import sys
 
+if "CRUNCH_BUILD_DIR" in os.environ.keys():
+    build_dir = os.environ["CRUNCH_BUILD_DIR"]
+else:
```
Comment (slipher) [inline test/test.py:9]: These 4 lines can be written as `build_dir = os.environ.get("CRUNCH_BUILD_DIR")`

File: test/test.py [hunk 2]
```diff
@@ -4,32 +4,62 @@
...
     if os.path.exists(windows_build_dir):
-        build_dir = windows_build_dir
+        return windows_build_dir
+    return build_dir
 
-    executable_path = os.path.join(build_dir, executable_name)
+def get_executable_path(executable_name):
+    if crunch_extension:
+        executable_extension = crunch_extension
+    elif sys.platform == "win32":
```
Comment (slipher) [inline test/test.py:52]: Not necessary in the native Windows case since `.exe` can be omitted from a command.

Comment (illwieckz) [reply test/test.py:52]: If I'm right the Windows CI failed without, the error was that the file was not found, Python is likely executing the binary directly without passing the command to a shell.

I also need the explicit extension to test that code on Linux with Python doing `wine crunch.exe …`.

File: test/test.py (hunk 1 above)
Comment (illwieckz) [reply test/test.py]: I used `os.getenv("CRUNCH_BUILD_DIR", "build")` instead.

File: test/test.py (hunk 2 above)
Comment (illwieckz) [reply test/test.py:41]: > I also need the explicit extension […]

Well, this use case use the environment variable…

I don't think it's bad to be explicit anyway.

Comment (illwieckz) [reply test/test.py]: So, it works without extension, and removing it also allows to delete multiple lines and a variable, so I remove that.

**Comment by illwieckz** [comment]
To avoid some warnings related to `strncpy` I imported a copy (without asserts) of `strcpy_safe` from `crn_strutils.cpp` into `example2.cpp` (which is expected to not build against `libcrn`).

I also added two warning fixes.

File: example2/example2.cpp [hunk 3]
```diff
@@ -28,11 +28,29 @@
 
 using namespace crnlib;
 
-#if defined(_WIN32)
```
[the remark this answers is not in the document]
Comment (illwieckz) [reply example2/example2.cpp]: `strncpy` doesn't guarantee to write an ending `\0`.

[the remark this answers is not in the document]
Comment (slipher) [reply example2/example2.cpp]: If the buffer is too small for the filename, it's not going to work regardless of what string copy function you used. How about exiting if the buffer is too small and then using memcpy?

[the remark this answers is not in the document]
Comment (illwieckz) [reply example2/example2.cpp]: I did that:

```c++
#if defined(_WIN32)
#define example2_strcpy_safe(d, l, s) strcpy_s(d, l, s)
#else
void example2_strcpy_safe(char *d, size_t l, const char* s) {l = strnlen(s, l - 1); memcpy(d, s, l); d[l] = '\0';}
#endif
```

I don't really mind the code not being optimal or about handling the case when the name is too long (it's still an example as you said), I just want the examples to be valid C++, portable, and not triggering reports for things known to go against best-practices.

**Comment by slipher** [comment]
LGTM

**Pull request merged**